utils/src/test/java/software/amazon/awssdk/utils/BinaryUtilsTest.java

```diff
@@ -159,4 +159,46 @@ public void testCopyBytesFrom_DirectByteBuffer_Idempotent() {
         assertTrue(partial1.length == 3);
         assertTrue(Arrays.equals(new byte[] {2, 3, 4}, partial1));
     }
+
+    @Test
+    public void testCopyRemainingBytesFrom_nullBuffer() {
+        assertNull(BinaryUtils.copyRemainingBytesFrom(null));
+    }
+
+    @Test
+    public void testCopyRemainingBytesFrom_noRemainingBytes() {
+        ByteBuffer bb = ByteBuffer.allocate(8);
+        bb.putInt(42);
+        bb.flip();
+
+        bb.getInt();
+
+        assertEquals(0, BinaryUtils.copyRemainingBytesFrom(bb).length);
+    }
+
+    @Test
+    public void testCopyRemainingBytesFrom_fullBuffer() {
+        ByteBuffer bb = ByteBuffer.allocate(8);
+        bb.putInt(42);
+        bb.putInt(42);
+        bb.flip();
+
+        byte[] copy = BinaryUtils.copyRemainingBytesFrom(bb);
+        assertEquals(bb, ByteBuffer.wrap(copy));
+        assertEquals(8, copy.length);
+    }
+
+    @Test
+    public void testCopyRemainingBytesFrom_partiallyReadBuffer() {
+        ByteBuffer bb = ByteBuffer.allocate(8);
+        bb.putInt(42);
+        bb.putInt(42);
+        bb.flip();
+
+        bb.getInt();
+
+        byte[] copy = BinaryUtils.copyRemainingBytesFrom(bb);
+        assertEquals(bb, ByteBuffer.wrap(copy));
+        assertEquals(4, copy.length);
+    }
```

Comment:
These tests might be clearer if we work with bytes instead of ints. Also, we should assert that the sequence is read correctly, eg: given a sequence of [ 1, 2, 3, 4 ] of which we read two bytes we can assert that we are left with a buffer that is equivalent to [ 3, 4 ] rather than just checking length.

Reply:
+1 bytes instead of ints. The `assertEquals()` should take care of the ordering check since `ByteBuffer#equals()` enforces that.

Reply:
Yes, sorry my comment wasn't clear. My concern about that was you were writing the same int (42) twice, therefore you don't know if what you are left with is the start or the end of the original buffer.

Reply:
Ah gotcha, yep that makes sense.